// storybook-addon-designs/esm/register/index.js
/** @jsx jsx */
import addons, { types } from '@storybook/addons';
import { jsx } from '@storybook/theming';
import { AddonName, PanelName, ParameterName } from '../addon';
import { Wrapper } from './containers/Wrapper';
export default function register(renderTarget) {
    addons.register(AddonName, function (api) {
        var title = 'Design';
        var render = function (_a) {
            var active = _a.active, key = _a.key;
            return (jsx(Wrapper, { key: key, channel: addons.getChannel(), api: api, active: active }));
        };
        if (renderTarget === 'tab') {
            addons.add(PanelName, {
                title: title,
                render: render,
                type: types.TAB,
                paramKey: ParameterName,
                route: function (_a) {
                    var storyId = _a.storyId;
                    return "/design/" + storyId;
                },
                match: function (_a) {
                    var viewMode = _a.viewMode;
                    return viewMode === 'design';
                }
            });
        }
        else {
            addons.addPanel(PanelName, { title: title, paramKey: ParameterName, render: render });
        }
    });
}
