// easy-react-form/modules/list.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.listContextPropType = exports.ListContext = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = List_;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _createReactContext = require('create-react-context');

var _createReactContext2 = _interopRequireDefault(_createReactContext);

var _form = require('./form');

var _ListPlugin = require('./plugins/ListPlugin.utility');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function List_(props) {
	return _react2.default.createElement(
		_form.Context.Consumer,
		null,
		function (context) {
			return _react2.default.createElement(List, _extends({}, props, { context: context }));
		}
	);
}

var List = function (_React$Component) {
	_inherits(List, _React$Component);

	function List(props) {
		_classCallCheck(this, List);

		var _this = _possibleConstructorReturn(this, (List.__proto__ || Object.getPrototypeOf(List)).call(this, props));

		_this.onReset = function () {
			_this.setState(_this.getInitialItemsState());
		};

		_this.getFieldName = function (i, name) {
			if (typeof i !== 'number') {
				throw new Error('Each `<Feild/>` in a `<List/>` must have an `i` property');
			}
			return (0, _ListPlugin.getFieldName)(_this.props.name, i, name);
		};

		_this.onRegisterField = function (name) {
			if (!_this.firstFieldName) {
				_this.firstFieldName = name;
			}
		};

		_this.add = function () {
			var context = _this.props.context;
			var _this$state = _this.state,
			    items = _this$state.items,
			    itemsCounter = _this$state.itemsCounter;

			_this.setState({
				itemsCounter: itemsCounter + 1,
				items: items.concat([itemsCounter])
			}, function () {
				var name = _this.props.name;

				if (_this.firstFieldName) {
					context.focus(name + ':' + itemsCounter + ':' + _this.firstFieldName);
				}
			});
		};

		_this.remove = function (i) {
			var items = _this.state.items;

			_this.setState({
				items: items.filter(function (_) {
					return _ !== i;
				})
			});
		};

		_this.map = function (func) {
			var items = _this.state.items;

			return items.map(function (i) {
				return func(i);
			});
		};

		_this.state = _extends({
			context: {
				getFieldName: _this.getFieldName,
				onRegisterField: _this.onRegisterField
			}
		}, _this.getInitialItemsState());
		return _this;
	}

	_createClass(List, [{
		key: 'componentDidMount',
		value: function componentDidMount() {
			var _props = this.props,
			    name = _props.name,
			    context = _props.context;

			context.onRegisterList(name, {
				onReset: this.onReset
			});
		}
	}, {
		key: 'getInitialItems',
		value: function getInitialItems() {
			var _props2 = this.props,
			    context = _props2.context,
			    name = _props2.name,
			    count = _props2.count;

			if (context.initialValues[name]) {
				return createIndexArray(context.initialValues[name].length);
			}
			return createIndexArray(count);
		}
	}, {
		key: 'getInitialItemsState',
		value: function getInitialItemsState() {
			var items = this.getInitialItems();
			return {
				items: items,
				itemsCounter: items.length
			};
		}
	}, {
		key: 'render',
		value: function render() {
			var _this2 = this;

			var children = this.props.children;
			var listContext = this.state.context;

			return _react2.default.createElement(
				_form.Context.Consumer,
				null,
				function (context) {
					return _react2.default.createElement(
						ListContext.Provider,
						{ value: listContext },
						children({
							map: _this2.map,
							add: _this2.add,
							remove: _this2.remove
						})
					);
				}
			);
		}
	}]);

	return List;
}(_react2.default.Component);

List.propTypes = {
	name: _propTypes2.default.string.isRequired,
	count: _propTypes2.default.number.isRequired,
	context: _propTypes2.default.object.isRequired,
	children: _propTypes2.default.func.isRequired
};

List.defaultProps = {
	count: 1
};

var ListContext = exports.ListContext = (0, _createReactContext2.default)();

function createIndexArray(size) {
	var array = new Array(size);
	var i = 0;
	while (i < size) {
		array[i] = i;
		i++;
	}
	return array;
}

var listContextPropType = exports.listContextPropType = _propTypes2.default.shape({
	getFieldName: _propTypes2.default.func.isRequired,
	onRegisterField: _propTypes2.default.func.isRequired
});
